cpu/samd5x/periph/can.c

```diff
@@ -0,0 +1,768 @@
+    dev->conf->can->CCCR.reg |= CAN_CCCR_DAR;
+
+    /* Reject all remote frames */
+    dev->conf->can->GFC.reg |= CAN_GFC_RRFE | CAN_GFC_RRFS;
+
+    /* Enable the peripheral's interrupt */
+    if (dev->conf->can == CAN0) {
+        NVIC_EnableIRQ(CAN0_IRQn);
+    }
+    else {
+        NVIC_EnableIRQ(CAN1_IRQn);
+    }
+
+    /* Enable reception interrupts: reception on FIFO0 and FIFO1 */
+    dev->conf->can->IE.reg |= CAN_IE_RF0NE | CAN_IE_RF1NE;
+    /* Enable transmission events interrupts */
+    dev->conf->can->IE.reg |= CAN_IE_TEFNE;
+    /* Enable the interrupt lines */
+    dev->conf->can->ILE.reg = CAN_ILE_EINT0 | CAN_ILE_EINT1;
+
+    _can = dev;
+
+    /* Exit initialization mode */
+    _exit_init_mode(dev->conf->can);
+
+    return res;
+}
+
+static uint8_t _form_message_marker(can_mm_t *can_mm)
+{
+    return can_mm->put | (can_mm->get << 4);
+}
+
+static int _send(candev_t *candev, const struct can_frame *frame)
+{
+    can_t *dev = container_of(candev, can_t, candev);
+
+    if (frame->can_dlc > CAN_MAX_DLEN) {
+        DEBUG_PUTS("CAN frame payload not supported");
+        return -1;
+    }
+
+    /* Check if the Tx FIFO is full */
+    if (dev->conf->can->TXFQS.reg & CAN_TXFQS_TFQF) {
+        DEBUG_PUTS("Tx FIFO is full");
+        return -1;
+    }
+
+    can_mm_t can_mm = {0};
+    uint8_t fifo_queue_put_idx = (dev->conf->can->TXFQS.reg >> CAN_TXFQS_TFQPI_Pos) & 0x1F;
+    DEBUG("Tx FIFO put index = %u\n", fifo_queue_put_idx);
+    uint8_t fifo_queue_get_idx = (dev->conf->can->TXFQS.reg >> CAN_TXFQS_TFGI_Pos) & 0x1F;
+    DEBUG("Tx FIFO get index = %u\n", fifo_queue_get_idx);
+    can_mm.put = fifo_queue_put_idx;
+    can_mm.get = fifo_queue_get_idx;
+
+    if (frame->can_id & CAN_EFF_FLAG) {
+        DEBUG_PUTS("Extended ID");
+        dev->msg_ram_conf.tx_fifo_queue[can_mm.put].TXBE_0.bit.ID = frame->can_id & CAN_EFF_MASK;
+        dev->msg_ram_conf.tx_fifo_queue[can_mm.put].TXBE_0.bit.XTD = 1;
+    }
+    else {
+        DEBUG_PUTS("Standard identifier");
+        dev->msg_ram_conf.tx_fifo_queue[can_mm.put].TXBE_0.bit.ID = (frame->can_id & CAN_SFF_MASK) << 18;
+        dev->msg_ram_conf.tx_fifo_queue[can_mm.put].TXBE_0.bit.XTD = 0;
+    }
+    dev->msg_ram_conf.tx_fifo_queue[can_mm.put].TXBE_0.bit.RTR = (frame->can_id & CAN_RTR_FLAG) >> 30;
+    dev->msg_ram_conf.tx_fifo_queue[can_mm.put].TXBE_0.bit.ESI = 1;
+    dev->msg_ram_conf.tx_fifo_queue[can_mm.put].TXBE_1.bit.DLC = frame->can_dlc;
+    dev->msg_ram_conf.tx_fifo_queue[can_mm.put].TXBE_1.bit.EFC = 1;
+    dev->msg_ram_conf.tx_fifo_queue[can_mm.put].TXBE_1.bit.MM = _form_message_marker(&can_mm);
+    memcpy((void *)dev->msg_ram_conf.tx_fifo_queue[can_mm.put].TXBE_DATA, frame->data, frame->can_dlc);
+
+    /* Request transmission */
+    dev->conf->can->TXBAR.reg |= (1 << can_mm.put);
+
+    return 0;
+}
+
+static bool _find_filter(can_t *can, const struct can_filter *filter, bool is_std_filter, int16_t *idx)
```

Comment:
why isn't idx just the return value 
```suggestion
static int _find_filter(can_t *can, const struct can_filter *filter, bool is_std_filter)
```
negative value -> not found

Reply:
that would also get these other `idx` from `uint16_t` into `int`

Reply:
also the attribute  `is_std_filter` is not needed as this information is allready part of filter
```suggestion
static bool _find_filter(can_t *can, const struct can_filter *filter)
```

Reply:
If I remove the boolean `is_std_filter` I will need to check whether the filter is standard or not at different places which means redundancy. I prefer to keep the check at just one place and store it there and use it when needed